[test] Try adding a `Violation::RULE` constant

ntBre opened the pull request:
Summary
--

This is a rough draft/experiment to associate a `Rule` with a `Violation` type for cheaper conversions between the two than the current approach of matching on the stringified names. It also moves `Violation` and `Diagnostic` and some related types out of `ruff_diagnostics` and into the linter to make the macro work.

This could definitely be a bad idea, but it seemed fun to try. Some of the module moves may actually be needed for the diagnostic refactor anyway. I'm mostly hoping to see a significant performance gain on codspeed.

Test Plan
--

Existing tests

ntBre commented:
Those benchmarks are pretty promising, maybe this is worth pursuing.

MichaReiser commented:
The speedups look promising. However, I don't think we want or even can store a `Rule` on the new `ruff_db::Diagnostic`. How would this work in that new model?

ntBre commented:
I think this is ready for review. I tried to regroup all of the imports I added to keep the `crate` blocks separate from the others.

I also changed `Message` to store a `NoqaCode` instead of a `Rule` directly, fortunately without losing much, if any, of the performance improvement. Some of the smaller benchmarks actually sped up a little, I think.

I also moved the `RULE` constant to a `ViolationMetadata::rule` method, replacing `ViolationMetadata::rule_name`, which was now unused.

ntBre commented on crates/ruff_linter/src/lib.rs:
```diff
@@ -14,12 +14,17 @@ pub use rule_selector::RuleSelector;
 pub use rules::pycodestyle::rules::IOError;
 
+pub use diagnostic::Diagnostic;
+pub(crate) use ruff_diagnostics::{Applicability, Edit, Fix, IsolationLevel, SourceMap};
```
This is not strictly necessary, it just made rewriting some of the imports automatically a little easier.

ntBre replied:
Definitely not, I just got a bit tired/lazy updating imports and wanted to replace all `ruff_diagnostics::*` imports with `crate::*` in this crate. It should be easy enough to remove the less common ones like `IsolationLevel` and `SourceMap`, though.

MichaReiser reviewed (approved):
Nice! 

The only thing that's unclear to me and I would like to get a short summary on what the next steps are is how you plan on removing `rule` (and `AsRule`) from `Diagnostic` because we won't be able to add it to `ruff_db::Diagnostic`. IMO, I think that's the next important refactor we shoul do

ntBre commented on crates/ruff_linter/src/diagnostic.rs:
```diff
@@ -3,30 +3,32 @@ use log::debug;
 }
 
 impl Diagnostic {
+    #[expect(clippy::needless_pass_by_value)]
```
Yes that's right. I was a bit confused why it wasn't already needless and thought further refactoring (like combining this with `Checker::report_diagnostic` in most cases) might make it not needless again. I'll add a note!

MichaReiser commented:
>  I was thinking that would take care of most uses, but the rule_enabled checks might be slightly trickier. 

Do you have a link to this check. I think that should be fine as long as you have a `ViolationMetadata`.

ntBre commented:
> Do you have a link to this check. I think that should be fine as long as you have a `ViolationMetadata`.

Oops, I meant `Checker::enabled`:

https://github.com/astral-sh/ruff/blob/8d5655a7baa1ecc84a906cbfc0e2294cf173dee7/crates/ruff_linter/src/checkers/ast/mod.rs#L454-L457

I think we can work around the uses that don't pass a `Rule` directly easily enough with a `ViolationMetadata`, like you said.

> It does sound right in that this is what I would hope to be possible. I don't have a clear enough picture to say whether it will be possible.

Sounds good. It's not entirely clear to me either, but it feels like it should work!

MichaReiser commented:
Why would we have to change `checker.enabled`? Most call sites should have access to the rule or are there call sites where we check if the diagnostic's rule is enabled?

ntBre commented:
Yes, there are call sites where `diagnostic.rule()` is passed. Some of them are already replaced in #18232, but there are still a few left.

MichaReiser commented:
> Did I really break the mypy primer? I didn't think I touched any ty files, but everything should be up to date and it looks okay on main. I also tried rerunning once. It looks like this is the root cause:

seems unlikely.

AlexWaygood commented:
I think it's possible that a recent commit on arviz (https://github.com/arviz-devs/arviz/commit/3205b82bb4d6097c31f7334d7ac51a6de37002d0? it landed a few hours ago) means that ty now crashes on that project when we didn't previously. There's the same crash being reported on https://github.com/astral-sh/ruff/pull/18333

AlexWaygood commented:
filed https://github.com/astral-sh/ruff/pull/18336 to fix

ntBre commented:
Hmm, I got a new cascade of errors on this one.

```
fatal[panic] Panicked at /home/runner/.cargo/git/checkouts/salsa-e6f3bb7c2a062968/4818b15/src/views.rs:114:9 when checking `/tmp/mypy_primer/projects/pegen/src/pegen/build.py`: `No downcaster registered for type `dyn ty_python_semantic::db::Db` in `Views``
fatal[panic] Panicked at /home/runner/.cargo/git/checkouts/salsa-e6f3bb7c2a062968/4818b15/src/zalsa.rs:210:32 when checking `/tmp/mypy_primer/projects/pegen/src/pegen/__main__.py`: `index `23` is uninitialized`
fatal[panic] Panicked at /home/runner/.cargo/git/checkouts/salsa-e6f3bb7c2a062968/4818b15/src/zalsa.rs:210:32 when checking `/tmp/mypy_primer/projects/pegen/src/pegen/parser.py`: `index `23` is uninitialized`
fatal[panic] Panicked at /home/runner/.cargo/git/checkouts/salsa-e6f3bb7c2a062968/4818b15/src/zalsa.rs:210:32 when checking `/tmp/mypy_primer/projects/pegen/src/pegen/grammar_parser.py`: `index `23` is uninitialized`
fatal[panic] Panicked at /home/runner/.cargo/git/checkouts/salsa-e6f3bb7c2a062968/4818b15/src/zalsa.rs:210:32 when checking `/tmp/mypy_primer/projects/pegen/src/pegen/grammar.py`: `index `23` is uninitialized`
fatal[panic] Panicked at /home/runner/.cargo/git/checkouts/salsa-e6f3bb7c2a062968/4818b15/src/zalsa.rs:210:32 when checking `/tmp/mypy_primer/projects/pegen/src/pegen/sccutils.py`: `index `23` is uninitialized`
fatal[panic] Panicked at /home/runner/.cargo/git/checkouts/salsa-e6f3bb7c2a062968/4818b15/src/zalsa.rs:210:32 when checking `/tmp/mypy_primer/projects/pegen/src/pegen/first_sets.py`: `index `23` is uninitialized`
fatal[panic] Panicked at /home/runner/.cargo/git/checkouts/salsa-e6f3bb7c2a062968/4818b15/src/zalsa.rs:210:32 when checking `/tmp/mypy_primer/projects/pegen/src/pegen/__init__.py`: `index `23` is uninitialized`
fatal[panic] Panicked at /home/runner/.cargo/git/checkouts/salsa-e6f3bb7c2a062968/4818b15/src/zalsa.rs:210:32 when checking `/tmp/mypy_primer/projects/pegen/src/pegen/tokenizer.py`: `index `23` is uninitialized`
fatal[panic] Panicked at /home/runner/.cargo/git/checkouts/salsa-e6f3bb7c2a062968/4818b15/src/zalsa.rs:210:32 when checking `/tmp/mypy_primer/projects/pegen/src/pegen/utils.py`: `index `23` is uninitialized`
fatal[panic] Panicked at /home/runner/.cargo/git/checkouts/salsa-e6f3bb7c2a062968/4818b15/src/zalsa.rs:210:32 when checking `/tmp/mypy_primer/projects/pegen/src/pegen/python_generator.py`: `index `23` is uninitialized`
fatal[panic] Panicked at /home/runner/.cargo/git/checkouts/salsa-e6f3bb7c2a062968/4818b15/src/zalsa.rs:210:32 when checking `/tmp/mypy_primer/projects/pegen/src/pegen/parser_generator.py`: `index `23` is uninitialized`
fatal[panic] Panicked at /home/runner/.cargo/git/checkouts/salsa-e6f3bb7c2a062968/4818b15/src/zalsa.rs:210:32 when checking `/tmp/mypy_primer/projects/pegen/src/pegen/grammar_visualizer.py`: `index `23` is uninitialized`
fatal[panic] Panicked at /home/runner/.cargo/git/checkouts/salsa-e6f3bb7c2a062968/4818b15/src/zalsa.rs:210:32 when checking `/tmp/mypy_primer/projects/pegen/src/pegen/web.py`: `index `23` is uninitialized`
fatal[panic] Panicked at /home/runner/.cargo/git/checkouts/salsa-e6f3bb7c2a062968/[481](https://github.com/astral-sh/ruff/actions/runs/15283313638/job/42987393320?pr=18234#step:6:482)8b15/src/zalsa.rs:210:32 when checking `/tmp/mypy_primer/projects/pegen/src/pegen/validator.py`: `index `23` is uninitialized
```

I'll try rerunning once to see if that helps.
